play-json/shared/src/test/scala/play/api/libs/json/ReadsSharedSpec.scala

```diff
@@ -58,6 +58,42 @@ class ReadsSharedSpec extends WordSpec with MustMatchers {
     }
   }
 
+  "Compose" should {
+    lazy val generated: Reads[Owner] = Json.reads
+
+    "preprocess a JSON object using a Reads" in {
+      implicit val reads: Reads[Owner] =
+        generated.composeWith(Reads[JsValue] {
+          case obj @ JsObject(_) => (obj \ "avatar").asOpt[String] match {
+            case Some(_) => JsSuccess(obj)
+            case _ => JsSuccess(obj + ("avatar" -> JsString("")))
+          }
+
+          case _ => JsError("Object expected")
+        })
+
+      Json.obj("login" -> "foo", "url" -> "url://id").
+        validate[Owner] mustEqual JsSuccess(Owner("foo", "", "url://id"))
+
+    }
+
+    "preprocess a JSON object using a function" in {
+      implicit val reads: Reads[Owner] = generated.compose {
+        val tx: PartialFunction[JsValue, JsValue] = {
```

Comment:
Out of interest do you need to supply the type of the `PartialFunction` here to satisfy the `compose` function signature or are you just writing the type out to be clear?

Reply:
`The argument types of an anonymous function must be fully known. (SLS 8.5)` as there are 2 `compose` functions (with the one deprecated for now)

Reply:
Updated @richdougherty